### PR TITLE
search: match unit fix for Unicode characters

### PR DESCRIPTION
**Pull request by egabancho**
Signed-off-by: Esteban J. G. Gabancho <esteban.gabancho@gmail.com>

### CODE REVIEW

File: invenio/modules/search/walkers/match_unit.py [hunk 1]
```diff
@@ -62,16 +63,21 @@ def match_unit(record, p, f=None, m='a', wl=None):
...
-    if m == 'e':
-        return str(record) == p
-    return p.search(str(record)) is not None
+
+    try:
+        if m == 'e':
+            return unicode(record) == p
+        return p.search(unicode(record)) is not None
+    except Exception, e:
+        import ipdb; ipdb.set_trace()
```
Comment (jalavik) [inline invenio/modules/search/walkers/match_unit.py]: ipdb :dancer:

Comment (egabancho) [inline invenio/modules/search/walkers/match_unit.py]: Ups!

File: invenio/utils/datastructures.py [hunk 2]
```diff
@@ -295,6 +295,9 @@ def __len__(self):
     def keys(self):
         return self._dict.keys()
 
+    def values(self):
```
Comment (jirikuncar) [inline invenio/utils/datastructures.py]: can you please separate this change to a new commit so we can backport it to maint-2.0?

Comment (egabancho) [inline invenio/utils/datastructures.py]: Should I split it in different branches or is fine like this?

Comment (jirikuncar) [inline invenio/utils/datastructures.py]: IMHO 3d6a650 should be backported to `maint-2.0`, hence new PR

**Comment by egabancho** [comment]
Depends on https://github.com/inveniosoftware/invenio/pull/3350

File: invenio/utils/datastructures.py (hunk 2 above)
Comment (egabancho) [inline invenio/utils/datastructures.py]: Done!

File: invenio/modules/search/walkers/match_unit.py [hunk 3]
```diff
@@ -62,16 +63,16 @@ def match_unit(record, p, f=None, m='a', wl=None):
...
                     for field in record])
-    elif isinstance(record, dict):
+    elif isinstance(record, MutableMapping):
         return any([match_unit(field, p, f=f, m=m, wl=wl)
                     for field in record.values()])
 
     if m == 'e':
-        return str(record) == p
-    return p.search(str(record)) is not None
+        return unicode(record) == p
```
Comment (jirikuncar) [inline invenio/modules/search/walkers/match_unit.py]: @egabancho ping - is it necessary?

Comment (egabancho) [inline invenio/modules/search/walkers/match_unit.py]: As we discuss IRL probably not (neither the `str`)

I will test it ASAP and update accordingly

Comment (jirikuncar) [inline invenio/modules/search/walkers/match_unit.py]: @egabancho ping

Comment (egabancho) [inline invenio/modules/search/walkers/match_unit.py]: I think we need it, record at this point could be something like `int`

File: invenio/modules/search/walkers/match_unit.py [hunk 4]
```diff
@@ -62,16 +63,16 @@ def match_unit(record, p, f=None, m='a', wl=None):
...
                     for field in record])
-    elif isinstance(record, dict):
+    elif isinstance(record, MutableMapping):
         return any([match_unit(field, p, f=f, m=m, wl=wl)
                     for field in record.values()])
 
     if m == 'e':
-        return str(record) == p
-    return p.search(str(record)) is not None
+        return record == p
```
Comment (jirikuncar) [inline invenio/modules/search/walkers/match_unit.py]: `six.text_type(record) == p`?

Comment (egabancho) [inline invenio/modules/search/walkers/match_unit.py]: I'm not sure about this one, is `p` always a string? ( I mean we are not comparing types here)

**Comment by egabancho** [comment]
ping @jirikuncar

**Pull request merged**